ranger/ranger.go

```diff
@@ -136,8 +140,111 @@ func (r *Ranger) Context() (context.Context, context.CancelFunc) { return r.ctx,
 func (r *Ranger) DB() postgres.DatabaseService                   { return r.db }
 func (r *Ranger) Env() trails.Environment                        { return r.env }
 func (r *Ranger) Metadata() Metadata                             { return r.metadata }
+func (r *Ranger) ReqParser() *req.Parser                         { return r.req }
 func (r *Ranger) SessionStore() session.SessionStorer            { return r.sessions }
 
+// ParseBody reads r.Body into payload and writes to w when this fails.
+//
+// payload must be a pointer to a struct in order receive r.Body's data.
+// Upon success, ParseBody returns true and payload's fields are hydrated with data from r.Body.
+// Upon failure, ParseBody writes to w and returns false.
+// The caller should not write to w when ParseBody fails.
+//
+// ParseBody expects r.Body to be JSON-encoded data.
+// If r.Body is incorrectly formatted JSON, ParseBody writes 400.
+//
+// ParseBody expects the decoded data to pass "validate" struct tags defined on payload's type.
+// If values in the data do not meet validation requirements, ParseBody writes 422.
+//
+// In all cases, if an unhandled error occurs, ParseBody writes 500.
+//
+// If control over response codes is desired, use ranger.ReqParser.ParseBody.
+func (ranger *Ranger) ParseBody(w http.ResponseWriter, r *http.Request, payload any) bool {
```

Comment:
I can see that, I left the method name broad since I do see this being the same signature for parsing/validating HTML forms (toggle around `Content-Type`). We've got just a handful of those on portal, so I wasn't too worried about including them here, but can walk that out if it seems helpful now.